### ISSUE TITLE
Unable to reset fibers

### ISSUE THREAD

**Issue by carme-hp** [maierbn/opendihu#9]
This issue is related to the need to be able to reset data when using the *FastMonodomainSolver*, together with the *MuscleContractionSolver* and we do implicit coupling with preCICE. This is related to #5, but in this [case](https://github.com/maierbn/opendihu/tree/debug-reset-fibers/examples/electrophysiology/fibers/cuboid_fast_monodomain) I consider a simpler dummy scenario for debugging purposes. 

I have created the following functions in the [FastMonodomainSolver](https://github.com/maierbn/opendihu/blob/debug-reset-fibers/core/src/specialized_solver/fast_monodomain_solver/fast_monodomain_solver_communication.tpp):
```
template<int nStates, int nAlgebraics, typename DiffusionTimeSteppingScheme>
void FastMonodomainSolverBase<nStates,nAlgebraics,DiffusionTimeSteppingScheme>::
loadFiberData(){
  LOG(INFO) << "in loadFiberData";
  LOG(INFO) << "vm values" << fiberData_[0].vmValues;
  fiberData_ = fiberDataOld_;
  LOG(INFO) << "new vm values" << fiberData_[0].vmValues;

}

template<int nStates, int nAlgebraics, typename DiffusionTimeSteppingScheme>
void FastMonodomainSolverBase<nStates,nAlgebraics,DiffusionTimeSteppingScheme>::
saveFiberData(){

  fiberDataOld_ = fiberData_;
}
```

Now, just for debugging, I've modified the time for loop in the [FastMonodomainSolver](https://github.com/maierbn/opendihu/blob/debug-reset-fibers/core/src/specialized_solver/fast_monodomain_solver/fast_monodomain_solver_compute.tpp) by adding calls to the previously defined functions. In every timestep I start by loading the saved data, do the computations as before and then save the data. I expected that this will make no difference to the results and so it seems.  

 I then added and if condition such that the data is saved only for t < end_time/2.  With this I expected that the diffusion will be stopped at t = end_time/2.  Here is what my modified time loop looks like:

```
  // loop over splitting time steps
  for (int timeStepNo = 0; timeStepNo < nTimeStepsSplitting_; timeStepNo++)
  {

    // load saved FiberData
    if (timeStepNo>0){
      loadFiberData();
      LOG(INFO) << "load fiber data ";
    }
    // perform Strang splitting

    double currentTime = startTime + timeStepNo * timeStepWidthSplitting;

    LOG(INFO) << "splitting " << timeStepNo << "/" << nTimeStepsSplitting_ << ", t: " << currentTime;

    // compute midTime once per step to reuse it. [currentTime, midTime=currentTime+0.5*timeStepWidth, currentTime+timeStepWidth]
    double midTime = currentTime + 0.5 * timeStepWidthSplitting;
    bool storeAlgebraicsForTransfer = timeStepNo == nTimeStepsSplitting_-1;   // after the last timestep, store the algebraics for transfer

    // perform splitting
    compute0D(currentTime, dt0D, nTimeSteps0D, false);
    compute1D(currentTime, dt1D, nTimeSteps1D, prefactor);
    compute0D(midTime,     dt0D, nTimeSteps0D, storeAlgebraicsForTransfer);

    updateFiberData();

    // save FiberData
    if (timeStepNo < nTimeStepsSplitting_/2){
      saveFiberData();
      LOG(INFO) << "save fiber data";
    }

    callOutputWriter(timeStepNo, currentTime, timeStepOutputInterval);

  }
```

What I observe in the results is that the diffusion continues and that I end up with the same results I had without the if condition. However if I look at the terminal output for t>end_time/2 I see that I'm indeed loading some old values:

```
in loadFiberData
vm values[-79.4501,-58.3737,19.7987,-24.6463,-72.9346,-84.3031,-84.397,-83.1819,-82.1121,-80.9372,-79.7931,-78.7198,-77.754,-76.9158,-76.216,-75.6705,-75.3508,-75.2946,-75.3508,-75.6705,-76.216,-76.9158,-77.754,-78.7198,-79.7931,-80.9372,-82.1121,-83.1818,-84.3971,-84.3026,-72.9348,-24.6424,19.8062,-59.2699,-77.2792,-74.5995,-75.0394,-74.9895,-74.9888,-74.9925]
new vm values[-74.989,-74.9891,-74.9884,-74.9926,-74.97,-75.0806,-74.5821,-76.6283,-68.7529,31.6883,-6.92646,-52.52,-83.6626,-84.79,-83.485,-82.4637,-81.597,-81.4773,-81.597,-82.4637,-83.485,-84.79,-83.6626,-52.5202,-6.92674,31.6881,-68.7527,-76.6283,-74.582,-75.0805,-74.97,-74.9926,-74.9884,-74.9891,-74.989,-74.989,-74.989,-74.989,-74.989,-74.989]
load fiber data 
splitting 19996/20000, t: 19.996
callOutputWriter with timeStepNo, currentTime = 19996 19.996
in loadFiberData
vm values[-79.4488,-58.3449,19.7697,-24.68,-72.9727,-84.3045,-84.3959,-83.1812,-82.1112,-80.9364,-79.7923,-78.7191,-77.7534,-76.9153,-76.2156,-75.6702,-75.3505,-75.2944,-75.3505,-75.6702,-76.2156,-76.9153,-77.7534,-78.7191,-79.7923,-80.9364,-82.1112,-83.1812,-84.396,-84.304,-72.9728,-24.6761,19.7771,-59.2435,-77.2788,-74.5999,-75.0393,-74.9895,-74.9888,-74.9925]
new vm values[-74.989,-74.9891,-74.9884,-74.9926,-74.97,-75.0806,-74.5821,-76.6283,-68.7529,31.6883,-6.92646,-52.52,-83.6626,-84.79,-83.485,-82.4637,-81.597,-81.4773,-81.597,-82.4637,-83.485,-84.79,-83.6626,-52.5202,-6.92674,31.6881,-68.7527,-76.6283,-74.582,-75.0805,-74.97,-74.9926,-74.9884,-74.9891,-74.989,-74.989,-74.989,-74.989,-74.989,-74.989]
load fiber data 
splitting 19997/20000, t: 19.997
callOutputWriter with timeStepNo, currentTime = 19997 19.997
in loadFiberData
vm values[-79.4475,-58.3159,19.7406,-24.7137,-73.0107,-84.3059,-84.3948,-83.1805,-82.1103,-80.9355,-79.7915,-78.7183,-77.7527,-76.9147,-76.2151,-75.6698,-75.3502,-75.2941,-75.3502,-75.6698,-76.2151,-76.9147,-77.7527,-78.7183,-79.7915,-80.9355,-82.1103,-83.1805,-84.3949,-84.3054,-73.0108,-24.7098,19.748,-59.2169,-77.2783,-74.6004,-75.0391,-74.9896,-74.9888,-74.9925]
new vm values[-74.989,-74.9891,-74.9884,-74.9926,-74.97,-75.0806,-74.5821,-76.6283,-68.7529,31.6883,-6.92646,-52.52,-83.6626,-84.79,-83.485,-82.4637,-81.597,-81.4773,-81.597,-82.4637,-83.485,-84.79,-83.6626,-52.5202,-6.92674,31.6881,-68.7527,-76.6283,-74.582,-75.0805,-74.97,-74.9926,-74.9884,-74.9891,-74.989,-74.989,-74.989,-74.989,-74.989,-74.989]
load fiber data 
splitting 19998/20000, t: 19.998
callOutputWriter with timeStepNo, currentTime = 19998 19.998
in loadFiberData
vm values[-79.4462,-58.2869,19.7116,-24.7475,-73.0486,-84.3073,-84.3937,-83.1798,-82.1094,-80.9347,-79.7906,-78.7176,-77.7521,-76.9142,-76.2147,-75.6695,-75.35,-75.2938,-75.35,-75.6695,-76.2147,-76.9142,-77.7521,-78.7176,-79.7906,-80.9347,-82.1094,-83.1798,-84.3938,-84.3068,-73.0488,-24.7435,19.7189,-59.1903,-77.2779,-74.6008,-75.0389,-74.9896,-74.9888,-74.9925]
new vm values[-74.989,-74.9891,-74.9884,-74.9926,-74.97,-75.0806,-74.5821,-76.6283,-68.7529,31.6883,-6.92646,-52.52,-83.6626,-84.79,-83.485,-82.4637,-81.597,-81.4773,-81.597,-82.4637,-83.485,-84.79,-83.6626,-52.5202,-6.92674,31.6881,-68.7527,-76.6283,-74.582,-75.0805,-74.97,-74.9926,-74.9884,-74.9891,-74.989,-74.989,-74.989,-74.989,-74.989,-74.989]
load fiber data 
splitting 19999/20000, t: 19.999
callOutputWriter with timeStepNo, currentTime = 19999 19.999
File "solver_structure.txt" written.
Wrote 4 stimulation times to file "out/stimulation.log".
Total user time: 6s
File "out/mappings_between_meshes.txt" written.
```

I don't know how to explain that the fiber diffusion continues eve thought I load the saved values. I would expect that fort>end_time/2 we would always calculate the same results that we would then rewritten.  Is `DataFibers_` the data structure that is actually used for the computations? 

** How to reproduce:**
 In branch [debug-reset-fibers(https://github.com/maierbn/opendihu/tree/debug-reset-fibers) go to the  [case](https://github.com/maierbn/opendihu/tree/debug-reset-fibers/examples/electrophysiology/fibers/cuboid_fast_monodomain) and run:
` ./fastmonodomain ../settings_fastmonodomain.py`

> **Note**
> Since we are running the *FastMonodomainSolver* by itself (not coupled to another solver), the callOutputWriter is called at every timestep. This modification of the solver is discussed in #7 .

**Comment by maierbn** [comment]
Thanks for the detailed report, I can see what went wrong. The variable `fiberData_` in `FastMonodomainSolver` is only used as a buffer for inter-process communication. The actual values (the so-called "states" of the differential-algebraic equation) are stored in `fiberPointBuffers_` during computation. I fixed it in PR #10.

As you know, the whole purpose of this  `FastMonodomainSolver` compared to only using the plain Strang splitting is to compute action potential propagation (0D-1D) as fast as possible (That's how we got the 180 GFlop/s or 25 % peak performance on the Cascade Lake </end of ad break> ;-). 
* The 0D model is solved by Heun's method. Given as a CellML model, an explicitly vectorized code snippet for one Heun step is generated at runtime, compiled and loaded such that constants can be inlined and everything is AVX-2 or AVX-512. 
* The 1D model is solved within the opendihu core by implicit Euler. Since it is a tri-diagonal system, Thomas algorithm is used inplace. No data is copied between 0D, 1D and 0D computations (one Strang step is 0D-1D-0D), everything resides in `fiberPointBuffers_`.
* The domain decomposition (dd) for the 3D and 0D/1D computations is different.
![image](https://user-images.githubusercontent.com/23456394/217923061-db6a05de-6cb8-4c6e-a4a6-a412fddab8a0.png)
  * The left two images show the dd (4 processors) while the 3D model is solved and during the data mapping 3D->0D. The rationale is to use the same dd for the 3D and 0D models because we need to map between the meshes. The 3D dd is also the one that is created locally as specified from the settings file. 
  * The right two images show the dd in shapes of long stripes, while the 0D and 1D models get solved. It is required to have all data for a 1D model on the same process to do the Thomas algorithm without communication.
* Therefore, we need to communicate data from 3D to 0D/1D right before the Strang splitting starts and back from 0D/1D to 3D after the specified number of Strang splittings are completed:
![image](https://user-images.githubusercontent.com/23456394/217929688-0879562d-1d4c-4e30-8e66-af3bf95fb65e.png)
* The MPI communication uses the buffers `fiberData_`. The function `fetchFiberData()` first copies the relevant states (not all!) from `fiberPointBuffers_` → `fiberData_`, does the communication (`MPI_Scatterv`), and copied the received values back to `fiberPointBuffers_`. The function `updateFiberData()` does the same in reverse.

I saw that you moved `updateFiberData()` into the Strang loop https://github.com/maierbn/opendihu/blob/99a166448f61bfa535d45fc971a276fda50f08f5/core/src/specialized_solver/fast_monodomain_solver/fast_monodomain_solver_compute.tpp#L126. This does not make sense and breaks parallel execution (I think). 

Also, moving the output writer inside the loop only works for the serial case, as the dd has changed. If you want more output files, then don't use FastMonodomainSolver but the normal nested solver structure (Should work with the same settings file, maybe minor tweaking). Or you could limit the number of Strang splitting iterations inside the `FastMonodomainSolver` to 1 and wrap it in a `RepeatedCall`.
But anyhow nice that you helped yourself and, just for debugging, got more output.

One small note on the style guide: The opening brace `{` has its one line everywhere in the code and all members and methods have a javadoc comment.

**Comment by carme-hp** [comment]
Hi, thanks a lot for your detailed explanation, it was extremely helpful. 

Thanks to the changes suggested on your pull request I was able to save and load checkpoints for the fibers as expected (not only in the dummy scenario, but also when using the precice surface adapter - see [commit](https://github.com/maierbn/opendihu/commit/ee425e07b8a6fcadb1de97a06fb85207de7deab5)).

Just maybe one comment to your suggested solution in #10, instead of 
```
template<int nStates, int nAlgebraics, typename DiffusionTimeSteppingScheme>
void FastMonodomainSolverBase<nStates,nAlgebraics,DiffusionTimeSteppingScheme>::
restoreFiberDataCheckpoint()
{
  fiberPointBuffers_ = fiberPointBuffersLastCheckpoint_;
}
```
I had to use 
```
template<int nStates, int nAlgebraics, typename DiffusionTimeSteppingScheme>
void FastMonodomainSolverBase<nStates,nAlgebraics,DiffusionTimeSteppingScheme>::
restoreFiberDataCheckpoint()
{
  fiberPointBuffers_ = fiberPointBuffersLastCheckpoint_;
  updateFiberData();
}
```
